ompi/mca/coll/ucc/coll_ucc_reduce_scatter_block.c

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright (c) 2021 Mellanox Technologies. All rights reserved.
+ * Copyright (c) 2022 NVIDIA Corporation. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ */
+
+#include "coll_ucc_common.h"
+
+static inline
+ucc_status_t mca_coll_ucc_reduce_scatter_block_init(const void *sbuf, void *rbuf,
+                                                    int rcount,
+                                                    struct ompi_datatype_t *dtype,
+                                                    struct ompi_op_t *op,
+                                                    mca_coll_ucc_module_t *ucc_module,
+                                                    ucc_coll_req_h *req,
+                                                    mca_coll_ucc_req_t *coll_req)
+{
+    ucc_datatype_t         ucc_dt;
+    ucc_reduction_op_t     ucc_op;
+    int comm_size = ompi_comm_size(ucc_module->comm);
+
+    if (MPI_IN_PLACE == sbuf) {
+        /* TODO: UCC defines inplace differently:
+           data in rbuf of rank R is shifted by R * rcount */
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+    ucc_dt = ompi_dtype_to_ucc_dtype(dtype);
+    ucc_op = ompi_op_to_ucc_op(op);
+    if (OPAL_UNLIKELY(COLL_UCC_DT_UNSUPPORTED == ucc_dt)) {
+        UCC_VERBOSE(5, "ompi_datatype is not supported: dtype = %s",
+                    dtype->super.name);
+        goto fallback;
+    }
+    if (OPAL_UNLIKELY(COLL_UCC_OP_UNSUPPORTED == ucc_op)) {
+        UCC_VERBOSE(5, "ompi_op is not supported: op = %s",
+                    op->o_name);
+        goto fallback;
+    }
+    ucc_coll_args_t coll = {
+        .mask      = 0,
+        .coll_type = UCC_COLL_TYPE_REDUCE_SCATTER,
+        .src.info = {
+            .buffer   = (void*)sbuf,
+            .count    = rcount * comm_size,
```

Comment:
rcount * comm_size -> ((size_t)rcount) * comm_size to avoid potential INT overrun